manual_step/README

```diff
@@ -0,0 +1,25 @@
+# Ideas
+
+## Entry in jupyter notebook cell
+* Markup is generated in the cell corresponding to the RDF description of the manual step.
+* A "results" input section of the cell allows user to input URI of expected/resultant data sets.
+* Once output of manual task is specified, notebook may continue execution.
+
+## CWL tool
+* Optionally, command line prompt, or GUI prompt?
+* Contact an expert? Several ways:
+    * Contact a server, which processes the request and contacts the relevant expert, displaying the steps required, and giving them a graphical interface in which the desired results may be entered.
+    * Emails the specified expert directly, with the instructions. Expert may reply to the email.
+
+
+Test CWL manual tool with
+```cwl-runner --preserve-entire-environment test_workflow.cwl --manual_mode commandline```
+
+## Define manual step class in CWL spec?
+* Include RDF description of manual steps
+* Should be blocking? If the manual step depends on output of a previous step, then workflow needs to "wait" in some way until the manual task can be carried out.
```

Comment:
We have to think about how the workflow should remember that the latest run wasn't done. Might be as simple as storing a json describing when the last run was and until which step it was executed.

If it takes a lot of time to do a manual step it might be tricky to leave the jupyter kernel running the whole time or keep the prompt open.

This means the results of the previous steps can't just be stored in memory, they also have to be stored on disk.

Reply:
I was hoping to store the status of that along with the rest of the notebook data. The outcome of a manual step is a URI to e.g. a nanopub that indicates what was done, when, how and where any results have been stored. If such a URI is entered in the manual step's UI, then we know it has been "run". If empty, the notebook does not run any further until a success URI is provided. Once one has been provided, we could perhaps autogenerate a little "report" of the contents of that nanopub, that could be displayed below the relevant manual step cell.

Reply:
Sounds good, though I think we should also consider to store this information outside of the notebook (maybe both in the notebook and outside). I'm thinking of a file like you have in IDE's, the ones you always add to .gitignore. These files contain all kinds of information that are specific to the project structure.

If we store this information outside of the jupyter notebook it will also be easier to use this information in a different way.

Reply:
As I see it, the user might wish to open a fresh new notebook with the same protocol in it, then work through it again (perhaps with different people or data). That's why I'd rather have the manual step metadata stored within a given notebook's save file, rather than a file shared between all notebooks. But everything the user does will have been nanopublished anyway, so the only data being stored in the notebook is a reference to the different nanopubs that contain the real info, or of course potentially just a reference to locally stored data (can we "nanopublish" to local disk? That would make our code cleaner...)